core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -290,88 +312,113 @@ class AvroUpsertMonitorActor(val pollInterval: FiniteDuration,
-        publishMessageToUpdateImportStatus(
-          attributes.importId,
-          None,
-          ImportStatuses.Error,
-          Option(s"Failed to import data. The underlying error message is: ${t.getMessage}")
-        )
-        acknowledgeMessage(message.ackId)
-      } pipeTo self
+              }
+            case Failure(t) =>
+              val errMsg = t match {
+                case errRpt: RawlsExceptionWithErrorReport => errRpt.errorReport.message
+                case x                                     => x.getMessage
+              }
+              publishMessageToUpdateImportStatus(attributes.importId,
+                                                 Option(status),
+                                                 ImportStatuses.Error,
+                                                 Option(errMsg)
+              )
+          }
+        case Some(status) =>
+          logger.warn(
+            s"Received a double message delivery for import ID [${attributes.importId}] which is already in status [$status].  Acking message."
+          )
+          acknowledgeMessage(message.ackId)
+        case None =>
+          publishMessageToUpdateImportStatus(attributes.importId,
+                                             None,
+                                             ImportStatuses.Error,
+                                             Option("Import status not found")
+          )
+      }
+    } yield ()
+
+    // Make sure message is acknowledged in the case of any failure while trying to construct importFuture
+    importFuture.map(_ => ImportComplete) recover { case t =>
+      logger.error(s"unexpected error in importFuture for ${attributes.importId}: ${t.getMessage}", t)
+      publishMessageToUpdateImportStatus(
+        attributes.importId,
+        None,
+        ImportStatuses.Error,
+        Option(s"Failed to import data. The underlying error message is: ${t.getMessage}")
+      )
+      acknowledgeMessage(message.ackId)
+    } pipeTo self
+
+  }
+
+  private def lookupWorkspace(attributes: AvroUpsertAttributes): Future[Workspace] =
+    dataSource.inTransaction { dataAccess =>
+      // db query: the SQL to find the workspace in the db
+      // errorId: if the workspace is not found, what identifier do we put in the error message?
+      val (dbQuery, errorId) = attributes match {
+        case cwdsAttributes: CwdsUpsertAttributes =>
+          (dataAccess.workspaceQuery.findById(cwdsAttributes.workspaceId.toString), cwdsAttributes.workspaceId.toString)
+        case importServiceAttributes: ImportServiceUpsertAttributes =>
+          (dataAccess.workspaceQuery.findByName(importServiceAttributes.workspace),
+           importServiceAttributes.workspace.toString
+          )
+      }
+      dbQuery.map {
+        case Some(workspace) => workspace
+        case None =>
+          throw new RawlsException(
+            s"Workspace $errorId not found while importing entities"
+          )
+      }
+    }
```

Comment:
This `lookupWorkspace` method replaces the db query that previously existed in the AvroUpsertMonitorMessageParser, as well as the code that already existed for Import Service.

A problem in #2771 is that for cWDS messages, we would fire two db queries right after each other:
* look up the workspace by id, so we could populate the namespace/name into the attributes
* look up the workspace by namespace/name

Now, we only fire one db query.